existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,334 @@
+# Universal and Existential Quantification in Haskell
+
+In logic, there are two common symbols: universal quantifier and existential quantifier. 
+
+You might recognize them as $\forall$ (for all) and $\exists$ (there exists). 
+
+These symbols (and the concepts they stand for) are relevant to Haskellers as well, since both universal and existential quantification is possible in Haskell. 
+
+In this article, we will show how to explicitly write down both types of quantification via the `ExplicitForAll` extension. We'll show 
+you how can use it for things like reordering type variables, enabling other extensions, and even creating a heteregenous list in Haskell.
+
+## Universal quantification
+
+In Haskell, all type variables are universally quantified by default. It just happens implicitly. 
+
+To give an example, look at the `id` function.
+
+```haskell
+id :: a -> a
+```
+
+We can think of it as "for all types $a$, this function takes a value of that type and returns a value of the same type".
+
+With the `ExplicitForAll` extension, we can write that down explicitly. 
+
+```haskell
+{-# LANGUAGE ExplicitForAll #-}
+
+id :: forall a. a -> a    
+```
+
+### Examples of usage
+
+The syntax is simple – at the beginning of function or instance declaration, before any constraints or arguments are used, use
+the `forall` quantifier to introduce all type variables that you will use later.
+
+Here, we introduce four variables: `a`, `b`, `c`, and `m`:
+
+```Haskell
+func :: forall a b c m. a -> b -> m c -- OK, it compiles
+```
+
+All type variables that you introduce in your type signature should belong to exactly one `forall`. No more, no less.
+
+```haskell
+func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
+```
+
+Of course, we can also add constraints.
+For example, we might want to specify that the `m` variable needs an instance of the `Monad` typeclass.
+
+```Haskell
+func :: forall a b c m. Monad m => a -> b -> m c
+```
+So far, it might seem not very useful.
+Nothing changed when we added the quantifier because it was already there (although implicit).
+However, making the quantification explicit allows us to do some new things.
+
+### Practical use cases of universal quantification
+
+#### Reordering type variables
```

Comment:
Doesn't this also fall under supporting other extensions?

Reply:
Perhaps we can simplify it by just having a thesis of "universal quantification is necessary for other extensions" and then list the examples of where it is used in extensions. 

It doesn't change content, but the framing makes it much easier to understand what to do with it.